File: Data Structures & Algorithms/target-sum/submission-3.cpp
class Solution {
public:
    int count = 0;
    int findTargetSumWays(vector<int>& nums, int target) {
        dfs(0, nums, target);
        return count;
    }

    void dfs(int i, vector<int>& nums, int target){
        if(i==nums.size()){
            if(target==0) count++;
            return;
        }

        dfs(i+1, nums, target+nums[i]);
        dfs(i+1, nums, target-nums[i]);
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-3.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        curComb = []
        combs = []

        def helper(i, sum):
            if sum == target:
                combs.append(curComb.copy())
                return
            if i>=len(nums) or sum>target:
                return
            
            sum+=nums[i]
            curComb.append(nums[i])
            helper(i,sum)

            sum-=nums[i]
            curComb.pop()
            helper(i+1,sum)
        
        helper(0,0)
        return combs
